pkg/datapath/loader/compile.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2017-2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loader
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+)
+
+// OutputType determines the type to be generated by the compilation steps.
+type OutputType string
+
+const (
+	outputObject   = OutputType("obj")
+	outputAssembly = OutputType("asm")
+	outputSource   = OutputType("c")
+)
+
+// progInfo describes a program to be compiled with the expected output format
+type progInfo struct {
+	// Source is the program source (base) filename to be compiled
+	Source string
+	// Output is the expected (base) filename produced from the source
+	Output string
+	// OutputType to be created by LLVM
+	OutputType OutputType
+}
+
+// directoryInfo includes relevant directories for compilation and linking
+type directoryInfo struct {
+	// Library contains the library code to be used for compilation
+	Library string
+	// Runtime contains headers for compilation
+	Runtime string
+	// Output is the directory where the files will be stored
+	Output string
+}
+
+var (
+	compiler       = "clang"
```

Comment:
I don't want to block that PR with that, but maybe in future we should consider making that configurable. Users can have multiple versions of clang installed and sometimes in stable distros (RHEL, SLES etc.) the older versions are proposed as default ones. It would be nice to provide a possibility of picking a specific version (i.e. `clang6`).